File: packages/core/src/schemas/metadata/MetadataConstant.ts
import { IMetadataSchema } from "@typia/interface";

import { ClassProperties } from "../../typings/ClassProperties";
import { MetadataConstantValue } from "./MetadataConstantValue";

export class MetadataConstant {
  public readonly type: "boolean" | "bigint" | "number" | "string";
  public readonly values: MetadataConstantValue[];

  private constructor(props: ClassProperties<MetadataConstant>) {
    this.type = props.type;
    this.values = props.values.map(MetadataConstantValue.create);
  }

  public static create(
    props: ClassProperties<MetadataConstant>,
  ): MetadataConstant {
    return new MetadataConstant(props);
  }

  public static from(json: IMetadataSchema.IConstant): MetadataConstant {
    return MetadataConstant.create({
      type: json.type,
      values: json.values.map(MetadataConstantValue.from),
    });
  }

  public toJSON(): IMetadataSchema.IConstant {
    return {
      type: this.type,
      values: this.values.map((value) => value.toJSON()),
    };
  }
}

File: packages/core/src/schemas/metadata/MetadataConstant.go
package metadata

type MetadataConstant struct {
	Type   string
	Values []*MetadataConstantValue
}

type MetadataConstantProps struct {
	Type   string
	Values []*MetadataConstantValue
}

func CreateMetadataConstant(props MetadataConstantProps) *MetadataConstant {
	return &MetadataConstant{
		Type:   props.Type,
		Values: append([]*MetadataConstantValue(nil), props.Values...),
	}
}

func MetadataConstantFrom(json MetadataConstantJSON) *MetadataConstant {
	values := make([]*MetadataConstantValue, 0, len(json.Values))
	for _, value := range json.Values {
		values = append(values, MetadataConstantValueFrom(value))
	}
	return CreateMetadataConstant(MetadataConstantProps{
		Type:   json.Type,
		Values: values,
	})
}

func (m *MetadataConstant) ToJSON() MetadataConstantJSON {
	values := make([]MetadataConstantValueJSON, 0, len(m.Values))
	for _, value := range m.Values {
		values = append(values, value.ToJSON())
	}
	return MetadataConstantJSON{
		Type:   m.Type,
		Values: values,
	}
}
